### DISCUSSION TITLE
JPEG auto-rotate changes image size

### DISCUSSION THREAD

**Discussion by palexvs** [libvips/libvips#3806] [Q&A] [answered]
Hi, I'm trying to migrate from **ImageMagic** to **libvips** as it should be much faster. However, I have noticed that **libvips** often (for ~40% of my images) significantly changes (increases or decreases) an image size during auto-rotation when **ImageMagic** gives a much closer result.

I'm trying to understand why it happens and how to apply auto-rotation without loosing quality and gaining size.

Here are examples:
```
vips jpegsave orig.jpeg[autorot] -o vips.jpeg
convert -auto-orient orig.jpeg mini.jpeg
```

[example_1](https://github.com/libvips/libvips/assets/1415597/cbb9c444-9ed1-4e61-9fa9-f20ac7928060)
```
1.6M orig.jpeg
1.7M mini.jpeg
669K vips.jpeg
```

[example_2](https://github.com/libvips/libvips/assets/1415597/ca38f5ef-5115-448c-84a2-a7d93b7718e3)
```
357K orig.jpeg
357K mini.jpeg
417K vips.jpeg
```

```
convert -v
Version: ImageMagick 7.1.1-25 Q16-HDRI aarch64 21883 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(5.0) 
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg jxl lcms lqr ltdl lzma openexr png ps raw tiff webp xml zlib zstd
Compiler: gcc (4.2)
```

```
vips -v
vips-8.15.1
```

Thank you

**Comment by jcupitt** [comment, marked as answer]
Hi @palexvs,

I think this is because imagemagick tries to "guess" (from an examination of the huffman tables) the Q factor a JPEG was compressed with, and then by default will reencode with this same Q.

Using your example1:

```
$ vips copy example1.jpeg vips-90.jpg[Q=90]
$ vips copy example1.jpeg vips-75.jpg
$ convert example1.jpeg im.jpg
$ ls -l vips-* im.jpg example1.jpeg 
-rw-r--r-- 1 john john  780499 Jan  3 10:39 vips-75.jpg
-rw-r--r-- 1 john john 1738178 Jan  3 10:38 vips-90.jpg
-rw-rw-r-- 1 john john 1812478 Jan  3 10:38 im.jpg
-rw-rw-r-- 1 john john 1648867 Jan  3 10:37 example1.jpeg
```

So it looks like your original has been encoded with Q90. imagemagick tried to guess this setting and keeps the size about the same (very roughly). libvips uses the standard JPEG default Q factor of 75, unless overriden.

I think it's a difference in philosophy. libvips tries to be predictable and not do too much magic behind your back, whereas IM tries to guess what you want. That's great when it guesses right, but can be annoying and unexpected if it fails.
